web/packages/teleterm/src/mainProcess/contextMenus/tabContextMenu.ts

```diff
@@ -21,61 +21,160 @@ import {
   ipcRenderer,
   Menu,
   MenuItemConstructorOptions,
+  dialog,
 } from 'electron';
 
+import { ConfigService } from 'teleterm/services/config';
+import { Shell, makeCustomShellFromPath } from 'teleterm/mainProcess/shell';
+
+import type { Document } from 'teleterm/ui/services/workspacesService';
 import {
   TabContextMenuEventChannel,
   TabContextMenuEventType,
   TabContextMenuOptions,
 } from '../types';
 
-type MainTabContextMenuOptions = Pick<TabContextMenuOptions, 'documentKind'>;
+type MainTabContextMenuOptions = {
+  document: Document;
+};
 
-export function subscribeToTabContextMenuEvent(): void {
+export function subscribeToTabContextMenuEvent(
+  shells: Shell[],
+  configService: ConfigService
+): void {
   ipcMain.handle(
     TabContextMenuEventChannel,
     (event, options: MainTabContextMenuOptions) => {
       return new Promise(resolve => {
+        let preventAutoPromiseResolveOnMenuClose = false;
```

Comment:
This is fine for now, but in the future in similar situations I think it's better to provide a universal way of handling both types of actions.

One action type resolves to an event the moment that an item is clicked (and the menu is closed), the other action type needs to perform some extra stuff after a click and the closing of the menu. But we also need to support a situation where the menu is closed and no action is chosen.

This could be supported by something like in the example below. The advantage is that we're getting rid of a special boolean variable that affects how the whole function works, instead each action abides by the same rules.

```typescript
let currentAction = Promise.resolve(undefined);

const performAction = (thunk: () => Promise<Event> | Event) => {
  return () => {
    currentAction = Promise.resolve(thunk())
  }
}

// …

            {
              label: 'Close',
              click: performAction(() => ({ event: TabContextMenuEventType.Close })),
            },

// …

              {
                label: customShell
                  ? `Change Custom Shell (${customShell.friendlyName})…`
                  : 'Select Custom Shell…',
                click: performAction(async () => { …

// …

        ).popup({
          callback: () =>
            resolve(currentAction),
        });

```

Reply:
Yeah, I felt that there must be a better way to solve this, but honestly I didn't spend too much time on it and just came up with this quick fix.
Thanks for the example, I will keep that in mind (and I left a link in the comment above `preventAutoPromiseResolveOnMenuClose` so it doesn't get lost).